test/e2e/canary_test.go

```diff
@@ -0,0 +1,176 @@
+	}
+
+	// check that the default ingress controller is ready
+	def := &operatorv1.IngressController{}
+	if err := waitForIngressControllerCondition(t, kubeClient, 5*time.Minute, defaultName, defaultAvailableConditions...); err != nil {
+		t.Fatalf("failed to observe expected conditions: %v", err)
+	}
+
+	if err := kubeClient.Get(context.TODO(), defaultName, def); err != nil {
+		t.Fatalf("failed to get default ingresscontroller: %v", err)
+	}
+
+	// Get default ingress controller deployment
+	deployment := &appsv1.Deployment{}
+	if err := kubeClient.Get(context.TODO(), controller.RouterDeploymentName(def), deployment); err != nil {
+		t.Fatalf("failed to get ingresscontroller deployment: %v", err)
+	}
+
+	// Get canary route
+	canaryRoute := &routev1.Route{}
+	name := controller.CanaryRouteName()
+	err = wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		if err := kubeClient.Get(context.TODO(), name, canaryRoute); err != nil {
+			t.Logf("failed to get canary route %s: %v", name, err)
+			return false, nil
+		}
+
+		return true, nil
+	})
+
+	image := deployment.Spec.Template.Spec.Containers[0].Image
+	clientPod := buildCanaryCurlPod("canary-route-check", canaryRoute.Namespace, image, canaryRoute.Spec.Host)
+	if err := kubeClient.Create(context.TODO(), clientPod); err != nil {
+		t.Fatalf("failed to create pod %s/%s: %v", clientPod.Namespace, clientPod.Name, err)
+	}
+	defer func() {
+		if err := kubeClient.Delete(context.TODO(), clientPod); err != nil {
+			t.Fatalf("failed to delete pod %s/%s: %v", clientPod.Namespace, clientPod.Name, err)
+		}
+	}()
+
+	// Test canary route and verify that the hello-openshift echo pod is running properly.
+	err = wait.PollImmediate(1*time.Second, 5*time.Minute, func() (bool, error) {
+		readCloser, err := client.CoreV1().Pods(clientPod.Namespace).GetLogs(clientPod.Name, &corev1.PodLogOptions{
+			Container: "curl",
+			Follow:    false,
+		}).Stream(context.TODO())
+		if err != nil {
+			return false, nil
+		}
+		scanner := bufio.NewScanner(readCloser)
+		defer func() {
+			if err := readCloser.Close(); err != nil {
+				t.Errorf("failed to close reader for pod %s: %v", clientPod.Name, err)
+			}
+		}()
+		foundBody := false
+		foundRequestPortHeader := false
+		for scanner.Scan() {
+			line := scanner.Text()
+			if strings.Contains(line, "Hello OpenShift!") {
+				foundBody = true
+			}
+			if strings.Contains(strings.ToLower(line), "x-request-port:") {
+				foundRequestPortHeader = true
+			}
+			if foundBody && foundRequestPortHeader {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	if err != nil {
+		t.Fatalf("failed to observe the expected canary response body: %v", err)
+	}
+}
+
+// TestCanaryStatusCondition ensures that the default
+// ingress controller has the canary check succeeding
+// status condition, which is an indication that canary
+// checks are functioning as intended.
+func TestCanaryStatusCondition(t *testing.T) {
```

Comment:
I wonder if it would be cleaner to add `{Type: ingresscontroller.IngressControllerCanaryCheckSuccessConditionType, Status: operatorv1.ConditionTrue}` to `defaultAvailableConditions` so that `TestDefaultIngressControllerSteadyConditions` would cover it.  

We'd need to add a `nondefaultAvailableConditions` or similar without `ingresscontroller.IngressControllerCanaryCheckSuccessConditionType` and use that in the tests that create custom ingresscontrollers, so something like this:

```go
var (
	nondefaultAvailableConditions = []operatorv1.OperatorCondition{
		{Type: operatorv1.IngressControllerAvailableConditionType, Status: operatorv1.ConditionTrue},
		{Type: operatorv1.LoadBalancerManagedIngressConditionType, Status: operatorv1.ConditionTrue},
		{Type: operatorv1.LoadBalancerReadyIngressConditionType, Status: operatorv1.ConditionTrue},
		{Type: operatorv1.DNSManagedIngressConditionType, Status: operatorv1.ConditionTrue},
		{Type: operatorv1.DNSReadyIngressConditionType, Status: operatorv1.ConditionTrue},
		{Type: ingresscontroller.IngressControllerAdmittedConditionType, Status: operatorv1.ConditionTrue},
	}
	defaultAvailableConditions = append(nondefaultAvailableConditions,
		{Type: ingresscontroller.IngressControllerCanaryCheckSuccessConditionType, Status: operatorv1.ConditionTrue})
)
// ...
```
No change here:
```go
func TestDefaultIngressControllerSteadyConditions(t *testing.T) {
	if err := waitForIngressControllerCondition(t, kclient, 10*time.Second, defaultName, defaultAvailableConditions...); err != nil {
		t.Errorf("did not get expected conditions: %v", err)
	}
}
```
Use `nondefaultAvailableConditions` here:
```go
func TestUserDefinedIngressController(t *testing.T) {
	name := types.NamespacedName{Namespace: operatorNamespace, Name: "test"}
	ing := newLoadBalancerController(name, name.Name+"."+dnsConfig.Spec.BaseDomain)
	if err := kclient.Create(context.TODO(), ing); err != nil {
		t.Fatalf("failed to create ingresscontroller: %v", err)
	}
	defer assertIngressControllerDeleted(t, kclient, ing)

	if err := waitForIngressControllerCondition(t, kclient, 5*time.Minute, name, nondefaultAvailableConditions...); err != nil {
		t.Errorf("failed to observe expected conditions: %v", err)
	}
}
```

Reply:
In the interest of time, I've created https://bugzilla.redhat.com/show_bug.cgi?id=1903165.
I will follow up on this after this initial PR merges.